### PR TITLE
feat(pubsub): support pubsub

### PR DESCRIPTION
**Pull request by idonava**
Added google pubsub instrumentation

### CODE REVIEW

File: src/helpers/events.js [hunk 1]
```diff
@@ -1,4 +1,58 @@
...
+
+/**
+ * Create and initialize a new serverless event in the epsagon format.
+ * @param {string} resourceType resourceType name
+ * @param {string} name Event name
+ * @param {string} operation Operation name
+ * @param {string} origin Origin name (optional)
+ * @returns {Object} Object with dnsEvent and event start time.
+ */
+const initialEvent = (resourceType, name, operation, origin) => {
```
Comment (ranrib) [inline src/helpers/events.js]: initialize

File: src/events/dns.js [hunk 2]
```diff
@@ -1,12 +1,9 @@
-const uuid4 = require('uuid4');
 const dns = require('dns');
 const shimmer = require('shimmer');
 const utils = require('../utils.js');
 const tracer = require('../tracer.js');
-const serverlessEvent = require('../proto/event_pb.js');
 const eventInterface = require('../event.js');
-const errorCode = require('../proto/error_code_pb.js');
-const { isBlacklistURL } = require('.././helpers/events');
+const { isBlacklistURL, initializeEvent } = require('.././helpers/events');
```
Comment (galbash) [inline src/events/dns.js]: initializeEvent should be under `eventInterface`

File: src/events/dns.js [hunk 3]
```diff
@@ -170,10 +143,10 @@ function wrapDnsResolveFunction(original) {
         if (typeof arg2 === 'object') {
             options = arg2;
         }
-        const { dnsEvent, startTime } = initialDnsEvent(original.name);
+        const { event, startTime } = initializeEvent('dns', original.name, 'dns');
```
Comment (galbash) [inline src/events/dns.js]: I would still use `event: dnsEvent` - the word event is too generic and reading the code might become confusing

File: src/events/google_cloud.js [hunk 4]
```diff
@@ -79,6 +87,132 @@ function bigQueryWrapper(wrappedFunction) {
     };
 }
 
+const getLastSplittedString = (array) => {
```
Comment (galbash) [inline src/events/google_cloud.js]: getLastSplittedItem (doesn't have to be string). also get the delimiter as a param and maybe move to helpers

File: src/events/google_cloud.js [hunk 5]
```diff
@@ -79,6 +87,132 @@ function bigQueryWrapper(wrappedFunction) {
...
+};
+
+/**
+ * Wrap pubsub request function.
+ * @param {Function} original pubsub request function.
+ * @returns {Function} The wrapped function
+ */
+function wrapPubSubRequestFunction(original) {
+    return function internalPubSubRequestFunction(config, callback) {
+        if (!config || !config.method) {
```
Comment (galbash) [inline src/events/google_cloud.js]: can this happen? we are not even catching a faliure here

File: src/events/google_cloud.js [hunk 6]
```diff
@@ -79,6 +87,132 @@ function bigQueryWrapper(wrappedFunction) {
...
+            const pubsubProjectId = this.projectId;
+            const { event, startTime } = initializeEvent(
+                GOOGLE_CLOUD_TYPES.pubsub.name, pubsubProjectId, config.method
+            );
+            const requestFunctionThis = this;
+            const responsePromise = new Promise((resolve) => {
+                patchedCallback = (err, arg2, arg3) => {
+                    if ((!pubsubProjectId || pubsubProjectId === GOOGLE_CLOUD_TYPES
+                        .defaultProjectId) && !!requestFunctionThis.projectId) {
+                        event.getResource().setName(requestFunctionThis.projectId);
```
Comment (galbash) [inline src/events/google_cloud.js]: I think otherwise you sould set it to the default (or it might be empty?)

File: src/events/google_cloud.js [hunk 7]
```diff
@@ -79,6 +87,132 @@ function bigQueryWrapper(wrappedFunction) {
...
+                    }
+                    const callbackResponse = {};
+                    if (arg2) {
+                        if (config.method === 'publish') {
```
Comment (galbash) [inline src/events/google_cloud.js]: use `switch`

File: src/events/google_cloud.js [hunk 8]
```diff
@@ -79,6 +87,132 @@ function bigQueryWrapper(wrappedFunction) {
...
+                            const messageIds = arg2.messageIds ? arg2.messageIds : arg2;
+                            if (Array.isArray(messageIds) && messageIds.length) {
+                                callbackResponse.messageIds = messageIds;
+                                const [messageId] = messageIds;
```
Comment (galbash) [inline src/events/google_cloud.js]: this is not rigid enough. Either find a better way or always report messageIds

File: src/events/google_cloud.js [hunk 9]
```diff
@@ -79,6 +87,132 @@ function bigQueryWrapper(wrappedFunction) {
...
+                                callbackResponse.messageId = messageId;
+                            }
+                        } else if (config.method === 'createSubscription') {
+                            callbackResponse.subscription = arg2;
+                        } else if (config.method === 'deleteSubscription') {
+                            if (config.reqOpts && config.reqOpts.subscription) {
+                                callbackResponse.subscription = getLastSplittedString(
```
Comment (galbash) [inline src/events/google_cloud.js]: what are you splitting here? is it really crucial?

File: src/events/google_cloud.js [hunk 10]
```diff
@@ -79,6 +87,132 @@ function bigQueryWrapper(wrappedFunction) {
...
+                                    config.reqOpts.subscription
+                                );
+                            }
+                        } else if (config.method === 'deleteTopic') {
+                            if (config.reqOpts && config.reqOpts.topic) {
+                                callbackResponse.topic = getLastSplittedString(
```
Comment (galbash) [inline src/events/google_cloud.js]: same

File: src/events/google_cloud.js [hunk 11]
```diff
@@ -79,6 +87,132 @@ function bigQueryWrapper(wrappedFunction) {
...
+                            }
+                        } else if (config.method === 'createTopic') {
+                            callbackResponse.topic = arg2;
+                        }
+                    }
+                    finalizeEvent(event, startTime, err, callbackResponse);
+                    resolve();
+                    if (callback) {
+                        if (arg3) {
+                            callback(err, arg2, ...arg3);
```
Comment (galbash) [inline src/events/google_cloud.js]: you are invoking callback invalidly here. either get `(err, arg2, ...arg3)` as parameters, or use `(err, arg2, arg3)` to invoke (based on the real callback arguments).

File: src/events/google_cloud.js [hunk 12]
```diff
@@ -79,6 +87,132 @@ function bigQueryWrapper(wrappedFunction) {
...
+                                );
+                            }
+                        } else if (config.method === 'createTopic') {
+                            callbackResponse.topic = arg2;
+                        }
+                    }
+                    finalizeEvent(event, startTime, err, callbackResponse);
+                    resolve();
+                    if (callback) {
+                        if (arg3) {
```
Comment (galbash) [inline src/events/google_cloud.js]: this if is redundant. if arg3 is undefined just pass it as undefined.

File: src/events/google_cloud.js [hunk 13]
```diff
@@ -79,6 +87,132 @@ function bigQueryWrapper(wrappedFunction) {
...
+                            callback(err, arg2);
+                        }
+                    }
+                };
+            });
+            tracer.addEvent(event, responsePromise);
+        } catch (err) {
+            tracer.addException(err);
+        }
+        return original.apply(this, [config, patchedCallback]);
```
Comment (galbash) [inline src/events/google_cloud.js:156]: are you sure this API has no option to not pass a callback and return a promise instead? please verify that.

File: src/events/google_cloud.js [hunk 14]
```diff
@@ -79,6 +87,132 @@ function bigQueryWrapper(wrappedFunction) {
...
+ * Wrap pubsub request function.
+ * @param {Function} original pubsub request function.
+ * @returns {Function} The wrapped function
+ */
+function wrapPubSubPullFunction(original) {
+    return function internalPubSubPullFunction(request, options, callback) {
+        try {
+            let pubsubProjectId;
+            if (request && request.subscription) {
+                const subscriptionSplited = request.subscription.split('/');
```
Comment (galbash) [inline src/events/google_cloud.js:171]: you have a utility function for this

File: src/events/google_cloud.js [hunk 15]
```diff
@@ -79,6 +87,132 @@ function bigQueryWrapper(wrappedFunction) {
...
+                if (subscriptionSplited.length > 1) {
+                    [, pubsubProjectId] = subscriptionSplited;
+                }
+            }
+            const { event, startTime } = initializeEvent(
+                GOOGLE_CLOUD_TYPES.pubsub.name, pubsubProjectId, 'Pull'
+            );
+            const patchedCallback = (err, res) => {
```
Comment (galbash) [inline src/events/google_cloud.js]: you are not calling the original callback!

File: src/events/google_cloud.js [hunk 16]
```diff
@@ -79,6 +87,132 @@ function bigQueryWrapper(wrappedFunction) {
...
+                        return acc;
+                    }, []);
+                    callbackResponse.messageIds = messageIds;
+                }
+                finalizeEvent(event, startTime, err, callbackResponse);
+            };
+            let responsePromise;
+            // in case callback has given from the client.
+            if (callback) {
+                responsePromise = original.apply(this, [request, options, patchedCallback]);
```
Comment (galbash) [inline src/events/google_cloud.js]: no need for a return in this block?

File: src/events/google_cloud.js [hunk 17]
```diff
@@ -79,6 +87,132 @@ function bigQueryWrapper(wrappedFunction) {
...
+                patchedCallback(null, response);
+                return res;
+            }, (err) => {
+                patchedCallback(err, null);
+                throw err;
+            });
+        } catch (err) {
+            tracer.addException(err);
+        }
+        return original.apply(this, [request, options, callback]);
```
Comment (galbash) [inline src/events/google_cloud.js]: when are you reaching this line? I feel like you might be calling the original function at least twice

File: src/events/dns.js (hunk 2 above)
Comment (galbash) [inline src/events/dns.js]: I think you are not importing finalizeEvent

**Review by galbash — CHANGES_REQUESTED** [review]
Some comments

File: src/events/google_cloud.js (hunk 9 above)
Comment (idonava) [reply src/events/google_cloud.js]: I'm splitting the subscription name, to delete a subscription, the user use: `pubsub.subscription('subs').delete();`
But the subscription name here is "projects/pubsub-projectid/subscriptions/subs"

File: src/events/google_cloud.js (hunk 14 above)
Comment (idonava) [reply src/events/google_cloud.js:171]: yes, to get the last item. here I want to get the second item.

File: src/events/google_cloud.js (hunk 16 above)
Comment (idonava) [reply src/events/google_cloud.js]: yes, fixed

File: src/events/google_cloud.js (hunk 17 above)
Comment (idonava) [reply src/events/google_cloud.js]: you right, fixed.

File: src/events/dns.js (hunk 2 above)
Comment (idonava) [reply src/events/dns.js]: moved.

File: src/events/dns.js (hunk 3 above)
Comment (idonava) [reply src/events/dns.js]: changed to slsEvent: dnsEvent -> soon will be push.

File: src/event.js [hunk 18]
```diff
@@ -103,3 +105,51 @@ module.exports.addLabelToMetadata = function addLabelToMetadata(event, key, valu
...
+    const resource = new serverlessEvent.Resource([
+        name,
+        resourceType,
+        operation,
+    ]);
+    const slsEvent = new serverlessEvent.Event([
+        `${resourceType}-${uuid4()}`,
+        utils.createTimestampFromTime(startTime),
+        null,
+        origin || resourceType,
```
Comment (galbash) [inline src/event.js]: why?

File: src/events/google_cloud.js [hunk 19]
```diff
@@ -79,6 +86,135 @@ function bigQueryWrapper(wrappedFunction) {
...
+                GOOGLE_CLOUD_TYPES.pubsub.name, pubsubProjectId, config.method
+            );
+            const requestFunctionThis = this;
+            const responsePromise = new Promise((resolve) => {
+                patchedCallback = (err, arg2, ...arg3) => {
+                    if ((!pubsubProjectId || pubsubProjectId === GOOGLE_CLOUD_TYPES
+                        .defaultProjectId) && !!requestFunctionThis.projectId) {
+                        pubsubEvent.getResource().setName(requestFunctionThis.projectId);
+                    }
+                    const callbackResponse = {};
```
Comment (galbash) [inline src/events/google_cloud.js]: call this `responseMetadata` instead

File: src/events/google_cloud.js [hunk 20]
```diff
@@ -79,6 +86,135 @@ function bigQueryWrapper(wrappedFunction) {
...
+                    callback(err, res);
+                }
+            };
+            // in case callback has given from the client.
+            if (callback) {
+                let patchedCallbackWithPromise = callback;
+                const promise = new Promise((resolve) => {
+                    patchedCallbackWithPromise = (err, res) => {
+                        patchedCallback(err, res);
+                        resolve();
```
Comment (galbash) [inline src/events/google_cloud.js]: resolve should be called before the user callback.

File: src/events/google_cloud.js [hunk 21]
```diff
@@ -79,6 +86,135 @@ function bigQueryWrapper(wrappedFunction) {
...
+                patchedCallback(null, response);
+                return res;
+            }, (err) => {
+                patchedCallback(err, null);
+                throw err;
+            });
+        } catch (err) {
+            tracer.addException(err);
+        }
+        return original.apply(this, [request, options, callback]);
```
Comment (galbash) [inline src/events/google_cloud.js]: this flow is still buggy:
1. user reaches line 201
2. user reaches line 202 - exception is raised
3. user reaches line 212
4. user reaches line 214

you called the original twice

**Review by galbash — CHANGES_REQUESTED** [review]
still a few comments

File: src/event.js (hunk 18 above)
Comment (idonava) [reply src/event.js]: my mistake of understanding the Event flow, fixed.

File: src/events/google_cloud.js (hunk 19 above)
Comment (idonava) [reply src/events/google_cloud.js]: replaced.

File: src/events/google_cloud.js (hunk 20 above)
Comment (idonava) [reply src/events/google_cloud.js]: tnx for that.

File: src/events/google_cloud.js (hunk 21 above)
Comment (idonava) [reply src/events/google_cloud.js]: tnx! fixed.

File: src/events/google_cloud.js [hunk 22]
```diff
@@ -79,6 +86,142 @@ function bigQueryWrapper(wrappedFunction) {
...
+                        return acc;
+                    }, []);
+                    responseMetadata.receivedMessages = receivedMessages;
+                }
+                eventInterface.finalizeEvent(pubsubEvent, startTime, err, responseMetadata);
+                if (callback) {
+                    callback(err, res);
+                }
+            };
+            // in case callback has given from the client.
```
Comment (galbash) [inline src/events/google_cloud.js]: has been given / was given

File: src/events/google_cloud.js [hunk 23]
```diff
@@ -79,6 +86,142 @@ function bigQueryWrapper(wrappedFunction) {
...
+                    patchedCallbackWithPromise = (err, res) => {
+                        resolve();
+                        patchedCallback(err, res);
+                    };
+                });
+                tracer.addEvent(pubsubEvent, promise);
+                return original.apply(this, [request, options, patchedCallbackWithPromise]);
+            }
+            clientPromiseRequest = original.apply(this, [request, options, callback]);
+            tracer.addEvent(pubsubEvent, clientPromiseRequest);
```
Comment (galbash) [inline src/events/google_cloud.js:222]: you should add the promise after you did then on it here. do:
```
clientPromiseRequest = clientPromiseRequest.then(...)
tracer.addEvent(event, clientPromiseRequest)
```

File: src/events/google_cloud.js [hunk 24]
```diff
@@ -79,6 +86,142 @@ function bigQueryWrapper(wrappedFunction) {
...
+                patchedCallback(null, response);
+                return res;
+            }, (err) => {
+                patchedCallback(err, null);
+                throw err;
+            });
+        } catch (err) {
+            tracer.addException(err);
+        }
+        if (!clientPromiseRequest) {
```
Comment (galbash) [inline src/events/google_cloud.js]: I think you should put this code in the catch (that's the only way it can run), will be more readable

**Review by galbash — CHANGES_REQUESTED** [review]
few small comments

File: src/events/google_cloud.js (hunk 24 above)
Comment (idonava) [reply src/events/google_cloud.js]: If tracer.addEvent throwing an error, and clientPromiseRequest was called - I don't want to call the original function twice.

File: src/events/google_cloud.js [hunk 25]
```diff
@@ -79,6 +86,143 @@ function bigQueryWrapper(wrappedFunction) {
...
+                if (callback) {
+                    callback(err, res);
+                }
+            };
+            // in case callback was given from the client.
+            if (callback) {
+                let patchedCallbackWithPromise = callback;
+                const promise = new Promise((resolve) => {
+                    patchedCallbackWithPromise = (err, res) => {
+                        resolve();
```
Comment (galbash) [inline src/events/google_cloud.js]: not good enough, cause it should be *after* we update the event. you may need to split `patchedCallback`.

**Review by galbash — DISMISSED** [review]
1 last comment :)

File: src/events/google_cloud.js (hunk 25 above)
Comment (idonava) [reply src/events/google_cloud.js]: fixed.

**Review by henperez — APPROVED** [review]

**Pull request merged**